unison-cli/src/Unison/Codebase/Editor/HandleInput.hs

```diff
@@ -1820,52 +1823,84 @@ handleDependents hq = do
+
+-- | Handle a @push@ command.
+handlePushRemoteBranch0 ::
   forall m v.
   Applicative m =>
-  -- | The URL to push to. If missing, it is looked up in `.unisonConfig`.
+  -- | The repo to push to. If missing, it is looked up in `.unisonConfig`.
   Maybe WriteRemotePath ->
   -- | The local path to push. If relative, it's resolved relative to the current path (`cd`).
   Path' ->
   -- | The push behavior (whether the remote branch is required to be empty or non-empty).
   PushBehavior ->
-  -- | The sync mode. Unused as of 21-11-04, but do look for yourself.
   SyncMode.SyncMode ->
   Action' m v ()
-handlePushRemoteBranch mayRepo path pushBehavior syncMode = do
-  srcb <- do
-    currentPath' <- use LoopState.currentPath
-    getAt (Path.resolve currentPath' path)
+handlePushRemoteBranch0 mayRepo path pushBehavior syncMode = do
   unlessError do
     (repo, remotePath) <- maybe (resolveConfiguredGitUrl Push path) pure mayRepo
-    (cleanup, remoteRoot) <-
-      unsafeTime "Push viewRemoteBranch" do
-        withExceptT Output.GitError do
-          viewRemoteBranch (writeToRead repo, Nothing, Path.empty)
-    -- We don't merge `srcb` with the remote branch, we just replace it. This push will be rejected if this rewinds time or misses any new
-    -- updates in the remote branch that aren't in `srcb` already.
-    case Branch.modifyAtM remotePath (\remoteBranch -> if shouldPushTo remoteBranch then Just srcb else Nothing) remoteRoot of
-      Nothing -> lift do
-        eval (Eval cleanup)
-        respond (RefusedToPush pushBehavior)
-      Just newRemoteRoot -> do
-        unsafeTime "Push syncRemoteRootBranch" do
-          withExceptT Output.GitError do
-            syncRemoteRootBranch repo newRemoteRoot syncMode
-        lift do
-          eval (Eval cleanup)
-          respond Success
+    lift (handlePushRemoteBranch repo path syncMode (Just (remotePath, pushBehavior)))
+
+-- Internal helper that implements pushing to a remote repo, which generalizes @gist@ and @push@.
+handlePushRemoteBranch ::
+  forall m v.
+  Applicative m =>
+  -- | The repo to push to.
+  WriteRepo ->
+  -- | The local path to push. If relative, it's resolved relative to the current path (`cd`).
+  Path' ->
+  SyncMode.SyncMode ->
+  -- | The remote target. If missing, the given branch contents should be pushed to the repo repo without updating the
+  -- root namespace.
+  Maybe (Path, PushBehavior) ->
+  Action' m v ()
+handlePushRemoteBranch repo localPath syncMode remoteTarget = do
+  sourceBranch <- do
+    currentPath' <- use LoopState.currentPath
+    getAt (Path.resolve currentPath' localPath)
+
+  unlessError do
+    (cleanup, remoteRoot) <- viewRemoteBranch (writeToRead repo, Nothing, Path.empty) & withExceptT Output.GitError
+    (`finallyE` lift (eval (Eval cleanup))) do
+      case remoteTarget of
+        Nothing -> do
+          let opts = PushGitBranchOpts {setRoot = False, syncMode}
```

Comment:
Question: why don't we want `setRoot` here? If the remoteTarget is empty doesn't that mean we're pushing to the root?

Reply:
I didn't get it either. 😬

Reply:
Did you see the comment on the argument?

```haskell
  -- | The remote target. If missing, the given branch contents should be pushed to the repo repo without updating the
  -- root namespace.
  Maybe (Path, PushBehavior) ->
```

If that doesn't make it clear enough, I can reword

Reply:
Yes, but I didn't understand it at the time. Now it seems okay.

Wdyt about a little more self-describing type that's isomorphic to this?

Reply:
Sure, a throwaway type with better constructor names than `Nothing` or `Just` is always an option. In this case I'm not sure that's a win. Also, per the suggestion to allow a repo-less `gist` command by piggybacking on the existing default namespace key, I think the implementation will change soon after it lands anyway. The distinction between `gist` and `push` in this PR is basically that, in addition to the `syncRoot = True`/`syncRoot = False` thing, `push` digs around the `.unisonConfig` file whereas `gist` does not.